rust/src/applayerpgsql/parser.rs

```diff
@@ -0,0 +1,426 @@
+/* Copyright (C) 2021 Open Information Security Foundation
+ *
+ * You can copy, redistribute or modify this Program under the terms of
+ * the GNU General Public License version 2 as published by the Free
+ * Software Foundation.
+ *
+ * This program is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ * GNU General Public License for more details.
+ *
+ * You should have received a copy of the GNU General Public License
+ * version 2 along with this program; if not, write to the Free Software
+ * Foundation, Inc., 51 Franklin Street, Fifth Floor, Boston, MA
+ * 02110-1301, USA.
+ */
+
+use std::{fmt};
+use nom::combinator::rest;
+use nom::number::streaming::{be_u8, be_u16, be_u32};
+
+pub const PGSQL_DUMMY_PROTO_MAJOR: u16 = 1234; // 0x04d2
+pub const PGSQL_DUMMY_PROTO_MINOR_SSL: u16 = 5679; //0x162f
+pub const _PGSQL_DUMMY_PROTO_MINOR_GSSAPI: u16 = 5680; // 0x1630
+
+#[derive(Debug, PartialEq)]
+struct PgsqlParameter {
+    param_name: Vec<u8>,
+    param_value: Vec<u8>,
+}
+
+#[derive(Debug, PartialEq)]
+pub struct PgsqlStartupParameters {
+    user: PgsqlParameter,
+    database: Option<PgsqlParameter>,
+    // TODO size-wise, being Option<Vec> or just Vec here is the same. I wonder if 
```

Comment:
how did you reach this conclusion?

Reply:
As I was concerned about this, I searched and found an answer in StackOverflow that shared a macro where one could make such comparisons (here: https://stackoverflow.com/questions/16504643/what-is-the-overhead-of-rusts-option-type)
I used that macro to compare some possibilities, and that seemed to be the case, to me. Here are some of the comparisons I've tried: https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=076232646a400fa0fe16a2fe0d53917a

(longer story here: https://jufajardini.wordpress.com/2021/03/10/til-sizes-of-structs-with-option-fields-vs-normal-attributes/)

Reply:
`Vec<u8>` and `Option<Vec<u8>` do show the same size with `size_of`, but I think there is a piece missing.  `Vec<u8>` requires a `Vec::new()` which is going to use some more memory that doesn't appear to show up in `size_of`.

I wonder if a good rule of thumb for us would be:
- If under normal conditions, something will always be in the Vec, just initialize it.
- If its common enough that the list could be empty, wrap it in Option and allocate it as needed?

Reply:
Thanks for the clarification. I think that approach makes sense!
I was thinking yesterday that I could actually have something much simpler, like just one Vec where all parameters will be stored, after parsing, but not sure if that's good, and had some trouble in making that work with my current parsers. But if you'd consider that better, I'd find out how other implementations have done it, and make it work.

Reply:
In fact, considering memory allocation, having just one vec seems to make much more sense in my head, but as I'm still learning, I'll wait for more knowledgeable considerations x)

Reply:
While this is an interesting discussion, I would like to suggest to not worry about such things too much in this phase of the project. These kind of micro optimizations can easily add unnecessary complexity.

Reply:
Fair, and important advice, to someone like me :sweat_smile: I'll stay on the path with adding more tests, as you guided me to.